In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
data1 = pd.read_csv("/content/drive/MyDrive/롤 데이터 분석/data/lol_data1.csv")
data2 = pd.read_csv("/content/drive/MyDrive/롤 데이터 분석/data/lol_data2.csv")
data3 = pd.read_csv("/content/drive/MyDrive/롤 데이터 분석/data/lol_data3.csv")

In [3]:
data = pd.concat([data1,data2,data3,],axis=0)
data

tier rank  leaguePoints  champExperience  champLevel  championId  \
0       BRONZE    I            75            14444          15          78   
1       BRONZE    I            75             7904          11         201   
2       BRONZE    I            75            12153          14          25   
3       BRONZE    I            75             8014          11         412   
4       BRONZE    I            75            11402          13         115   
...        ...  ...           ...              ...         ...         ...   
62086  DIAMOND   IV            30             5125           9          78   
62087  DIAMOND   IV            30             9991          13         143   
62088  DIAMOND   IV            30             7181          10          16   
62089  DIAMOND   IV            30            14246          15          89   
62090  DIAMOND   IV            30            13284          15         893   

      championName  championTransform  commandPings  consumablesPurchased  \
0            Poppy                  0             2                     4   
1            Braum                  0             2                     6   
2          Morgana                  0             1                    10   
3           Thresh                  0             3                     5   
4            Ziggs                  0             1                     6   
...            ...                ...           ...                   ...   
62086        Poppy                  0             7                     5   
62087         Zyra                  0            11                     7   
62088       Soraka                  0             3                     5   
62089        Leona                  0            53                    15   
62090       Aurora                  0             0                     0   

       ...  turretsTakenWithRiftHerald  twentyMinionsIn3SecondsCount  \
0      ...                           0                             0   
1      ...                           0                             0   
2      ...                           0                             0   
3      ...                           5                             0   
4      ...                           0                             0   
...    ...                         ...                           ...   
62086  ...                           0                             0   
62087  ...                           0                             0   
62088  ...                           0                             0   
62089  ...                           0                             0   
62090  ...                           0                             0   

       twoWardsOneSweeperCount  unseenRecalls  \
0                            0              0   
1                            0              0   
2                            0              0   
3                            0              1   
4                            0              0   
...                        ...            ...   
62086                        0              0   
62087                        0              0   
62088                        0              0   
62089                        0              0   
62090                        0              0   

       visionScoreAdvantageLaneOpponent  visionScorePerMinute  \
0                              0.326022              0.683416   
1                             -0.420499              1.147305   
2                             -0.135613              1.542737   
3                             -0.278238              1.656584   
4                             -0.552488              0.912949   
...                                 ...                   ...   
62086                         -0.155352              1.396453   
62087                          0.148002              2.286809   
62088                          0.292397              1.697488   
62089                          0.134700     

# **변수 설명**

## **타겟**
* tier: 플레이어의 티어(랭킹 수준).
* rank: 플레이어의 랭크 (티어 내의 세부 순위).

## **선택 변수**##
* champExperience: 챔피언이 얻은 총 경험치.
* champLevel: 챔피언 레벨.
* championName: 챔피언 이름.
* commandPings: 팀원에게 보낸 명령 핑의 개수.
* consumablesPurchased: 구매한 소모품(포션, 와드 등)의 개수.
* damageDealtToBuildings: 건물에 가한 총 피해량.
* damageDealtToObjectives: 오브젝트(드래곤, 바론 등)에 가한 총 피해량.
* damageDealtToTurrets: 타워에 가한 총 피해량.
* damageSelfMitigated: 플레이어가 완화한 총 피해량.
* dangerPings: 위험을 알리는 핑의 개수.
* deaths: 사망 횟수.
* detectorWardsPlaced: 설치한 탐지 와드의 개수.
* doubleKills: 더블 킬(연속 두 명 처치) 횟수.
* dragonKills: 플레이어가 처치한 드래곤 수.
* enemyMissingPings: 적 위치가 보이지 않음을 알리는 핑 개수.
* enemyVisionPings: 적의 시야 위치를 알리는 핑 개수.
* firstBloodAssist: 첫 킬을 지원했는지 여부.
* firstBloodKill: 첫 킬을 기록했는지 여부.
* firstTowerAssist: 첫 타워 파괴에 기여했는지 여부.
* firstTowerKill: 첫 타워를 파괴했는지 여부.
* getBackPings: 돌아가라는 핑의 개수.
* goldEarned: 획득한 총 골드.
* goldSpent: 사용한 총 골드.
* holdPings: 대기하라는 핑의 개수.
* individualPosition: 플레이어가 게임 내에서 실제로 플레이한 포지션
* inhibitorKills: 직접 억제기를 파괴한 횟수
* inhibitorTakedowns: 억제기를 파괴에 기여한 횟수
* inhibitorsLost: 잃은 억제기 수.
* item0 - item6: 장비 슬롯에 장착된 아이템 ID.
* itemsPurchased: 구매한 아이템의 총 개수.
* killingSprees: 킬 연속 횟수.
* kills: 총 처치 횟수.
* largestCriticalStrike: 최대 치명타 피해량.
* largestKillingSpree: 연속 킬 최대 횟수.
* largestMultiKill: 최대 연속킬 수.
* longestTimeSpentLiving: 생존한 가장 긴 시간.
* magicDamageDealt: 가한 마법 피해량.
* magicDamageDealtToChampions: 챔피언에게 가한 마법 피해량.
* magicDamageTaken: 받은 마법 피해량.
* needVisionPings: 시야 요청 핑의 개수.
* neutralMinionsKilled: 처치한 중립 미니언 수.
* objectivesStolen: 훔친 오브젝트 수(예: 드래곤, 바론 등).
* objectivesStolenAssists: 오브젝트 스틸 지원 횟수.
* onMyWayPings : 플레이어의 가는중 핑 횟수.
* physicalDamageDealt: 가한 물리 피해량.
* physicalDamageDealtToChampions: 챔피언에게 가한 물리 피해량.
* physicalDamageTaken: 받은 물리 피해량.
* pushPings: 전진하라는 핑의 개수.
* sightWardsBoughtInGame: 게임 내에서 구매한 시야 와드의 개수.
* spell1Casts : Q스킬 사용 횟수
* spell2Casts : W스킬 사용 횟수
* spell3Casts : E스킬 사용 횟수
* spell4Casts : R스킬 사용 횟수
* summoner1Casts : D 주문 사용 횟수
* summoner2Casts : F 주문 사용 횟수
* summoner1Id : D 주문 종류
* summoner2Id : F 주문 종류
* timeCCingOthers: 다른 챔피언을 군중 제어(CC)한 총 시간.
* timePlayed: 플레이한 총 시간.
* totalAllyJungleMinionsKilled: 아군 정글에서 처치한 미니언 수.
* totalDamageDealt: 총 가한 피해량.
* totalDamageDealtToChampions: 챔피언에게 가한 총 피해량.
* totalDamageShieldedOnTeammates: 팀원에게 제공한 방어막의 총 피해량.
* totalDamageTaken: 받은 총 피해량.
* totalEnemyJungleMinionsKilled: 적 정글에서 처치한 미니언 수.
* totalHeal: 회복한 총 체력량.
* totalHealsOnTeammates: 팀원에게 제공한 총 치유량입니다. (힐링 스킬,아이템)
* totalMinionsKilled: 게임 내에서 플레이어가 처치한 미니언의 총 수.
* totalTimeCCDealt: 적에게 가한 총 군중 제어(CC)시간 적을 스턴, 슬로우 등으로 무력화한 시간(초 단위)
* totalTimeSpentDead: 게임 중 플레이어가 사망하여 대기한 총 시간(초 단위)
* totalUnitsHealed: 플레이어가 회복시킨 총 유닛 수. 자신과 팀원을 포함한 치유 대상 수.
* trueDamageDealt: 가한 순수 피해량입니다. 방어 무시 피해. 방어력이나 마법 저항에 영향 X
* trueDamageDealtToChampions: 챔피언에게 가한 순수 피해량
* trueDamageTaken: 받은 순수 피해량. 방어력이나 마법 저항과 무관하게 받은 피해량.
* turretKills: 플레이어가 파괴한 타워의 수.
* turretTakedowns: 타워 파괴에 기여한 횟수. 최종 타격을 하지 않아도 기여한 경우.
* turretsLost: 아군 팀이 잃은 타워의 수.
* visionClearedPings: 시야 제거와 관련된 핑의 수.주로 상대 시야를 제거할 때 사용됩니다.
* visionScore: 플레이어의 시야 점수입니다. 와드 설치, 제거, 적 시야 방해 등을 통해 얻은 점수입니다.
* visionWardsBoughtInGame: 게임 내에서 구매한 제어 와드의 수입니다.
* wardsKilled: 파괴한 적 와드의 수입니다.
* wardsPlaced: 설치한 와드의 수입니다.
* win: 게임의 승리 여부입니다. 승리 시 True, 패배 시 False로 표시됩니다.
* allInPings: 전투 시작 또는 돌격을 알리는 핑의 수입니다.
* assistMePings: 지원 요청 핑의 수입니다.
* assists: 어시스트 횟수입니다. 적을 처치하는 데 기여한 횟수를 나타냅니다.
* basicPings: 플레이어가 사용한 기본 핑의 수입니다.
* bountyLevel: 플레이어의 현상금 레벨입니다. 킬을 연속으로 성공할 때 부여되는 현상금 단계입니다.
* controlWardsPlaced: 설치한 제어 와드(핑 와드)의 총 개수입니다.
* damagePerMinute: 분당 평균 피해량을 나타냅니다. 게임 내 총 피해량을 플레이 시간으로 나눈 값입니다.
* damageTakenOnTeamPercentage: 팀 전체가 받은 피해량 중에서 플레이어가 받은 피해량의 비율을 나타냅니다.
* earlyLaningPhaseGoldExpAdvantage: 초반 라인전에서 얻은 골드 및 경험치 우위를 나타냅니다.
* effectiveHealAndShielding: 플레이어가 제공한 총 회복 및 보호막량을 나타냅니다. 팀원을 힐하거나 보호막을 준 양을 포함합니다.
* enemyChampionImmobilizations: 적 챔피언을 무력화(CC)한 횟수를 나타냅니다.
* fullTeamTakedown: 팀 전체가 참여한 처치 횟수를 나타냅니다.
* goldPerMinute: 분당 골드 획득량입니다. 게임에서 플레이어가 얻은 평균 골드 수입을 의미합니다.
* immobilizeAndKillWithAlly: 아군과 협력하여 적을 무력화(이동 불가 상태로 만듦)하고 처치한 횟수를 의미합니다.
* kTurretsDestroyedBeforePlatesFall: 포탑 방어막이 유지되는 시간 동안 파괴한 타워의 수입니다.
* kda: 킬-데스-어시스트(Kill-Death-Assist) 비율입니다. (킬 + 어시스트) / 데스 값으로 계산되며, 플레이어의 생존율과 기여도를 나타냅니다.
* killAfterHiddenWithAlly: 아군과 함께 적에게 숨어 있다가 처치한 횟수입니다.
* killParticipation: 팀의 총 킬에서 플레이어가 관여한 비율입니다.
* killedChampTookFullTeamDamageSurvived: 적 챔피언에게 팀의 총 피해를 가한 후 생존한 횟수입니다.
* landSkillShotsEarlyGame: 초반에 적중한 스킬 샷의 횟수입니다.
* laneMinionsFirst10Minutes: 게임 시작 후 10분 이내에 처치한 라인 미니언 수입니다.
* laningPhaseGoldExpAdvantage: 라인전에서 얻은 골드와 경험치 우위를 나타냅니다.
* outnumberedKills: 적이 수적으로 우세한 상황에서 처치한 적의 수입니다.
* soloKills: 단독 킬 횟수입니다. 아군의 도움 없이 적을 처치한 경우를 나타냅니다.
* stealthWardsPlaced: 은신 와드(일반 와드)를 설치한 횟수입니다.
* takedownsInEnemyFountain: 적의 포탑 안에서 적을 처치하거나 어시스트한 횟수입니다.
* teamBaronKills: 팀이 처치한 바론의 수입니다.
* teamDamagePercentage: 팀 전체 데미지에서 플레이어가 차지하는 비율입니다. 플레이어의 기여도를 나타냅니다.
* teamElderDragonKills: 팀이 처치한 장로 드래곤의 수입니다.
* teamRiftHeraldKills: 팀이 처치한 리프트 헤럴드의 수입니다.
* turretPlatesTaken: 파괴한 타워 방어막의 수입니다. 타워 방어막은 게임 초반에 타워에 부여되는 보호막입니다.
* visionScoreAdvantageLaneOpponent: 라인 상대와 비교했을 때 시야 점수 우위를 나타냅니다.
* visionScorePerMinute: 분당 평균 시야 점수입니다.
* wardTakedowns: 파괴한 와드의 수입니다.
* wardTakedownsBefore20M: 게임 시작 후 20분 이내에 파괴한 와드의 수입니다.
* wardsGuarded: 플레이어가 방어한 와드의 수입니다. 팀의 시야를 유지하거나 보호한 횟수로, 와드를 적의 공격으로부터 지켜낸 경우를 나타냅니다.



## **부가적인 확인이 필요한 변수**
* summonerLevel: 소환사 계정 레벨.
* lane: 플레이어가 주로 플레이한 라인(탑, 미드 등).
* role: 게임에서 수행한 역할(예: 서포트, 딜러 등).
* nexusLost: 잃은 넥서스 수.(게임이 항복으로 끝났는가, 질질 끌다가 넥서스가 부숴졌는가.)
* abilityUses: 플레이어가 스킬을 사용한 총 횟수입니다. 모든 스킬과 주문 사용 횟수를 포함하지 않기 때문일 가능성이 높습니다. abilityUses는 특정 조건 하에서만 주요 스킬 사용 횟수를 집계
* deathsByEnemyChamps: 적 챔피언에 의해 사망한 횟수입니다.
* dodgeSkillShotsSmallWindow: 작은 시간 내에 회피한 스킬 샷의 수를 나타냅니다.
* enemyJungleMonsterKills: 적 정글에서 처치한 몬스터 수입니다.
* firstTurretKilled: 첫 번째 타워를 파괴했는지 여부입니다.
* firstTurretKilledTime: 첫 번째 타워가 파괴된 시간(초 단위)입니다.
* gameLength: 게임의 총 길이(초 단위)입니다. 한 판이 끝날 때까지의 총 시간을 나타냅니다.(모든 소환사 동일)
* timePlayed: 플레이한 총 시간.(접속 끊김,도주 등 실제 플레이 시간 기록)
* jungleCsBefore10Minutes: 게임 시작 후 10분 이내에 처치한 정글 미니언 수(CS)입니다.
* maxCsAdvantageOnLaneOpponent: 라인 상대와 비교한 최대 CS(Creep Score) 우위입니다. 라인전에서 적 챔피언보다 더 많은 미니언을 처치하여 얻은 최대 우위를 나타냅니다.
* legendaryItemUsed: 전설 아이템을 사용한 횟수 또는 전설 아이템 사용 여부를 나타냅니다.
* maxKillDeficit: 게임 내에서 한 번에 가장 많이 뒤처졌던 킬 격차입니다. 적이 얼마나 많은 킬을 앞섰는지의 최대치를 나타냅니다.
* maxLevelLeadLaneOpponent: 라인 상대보다 앞선 최대 레벨 차이를 나타냅니다.
* skillshotsDodged: 적의 스킬 샷을 회피한 횟수입니다.
* skillshotsHit: 적에게 적중시킨 스킬 샷의 횟수입니다.
* survivedThreeImmobilizesInFight: 한 전투에서 세 번의 무력화(이동 불가) 효과를 받고도 생존한 횟수입니다.

## **삭제**
* leaguePoints: 리그 포인트(LP).
* championId: 챔피언의 고유 ID.
* championTransform: 챔피언이 변신했는지 여부.
* gameEndedInSurrender: 게임이 항복으로 종료되었는지 여부.
* nexusKills: 넥서스 막타를 쳐서 파괴했는가.
* nexusTakedowns: 넥서스를 공격한 횟수.
* participantId: 플레이어의 고유 ID.
* pentaKills: 펜타킬 횟수(연속 5명 처치).
* quadraKills: 쿼드라킬 횟수(연속 4명 처치).
* profileIcon: 플레이어가 사용한 프로필 아이콘 ID.
* teamId: 100(블루),200(레드)
* teamPosition: 캐릭터 배정 당시의 포지션
* tripleKills: 트리플 킬(연속 3명 처치) 횟수입니다.
* unrealKills: 언리얼 킬 횟수.빠른 시간 안에 5명 이상을 처치한 경우 기록.
* baronKills: 플레이어가 막타로 처치한 바론의 수입니다.
* baronTakedowns: 바론을 처치에 기여한 횟수
* blastConeOppositeOpponentCount: 적을 맞은 반대편으로 날려버린 블라스트 콘(폭발식 식물) 사용 횟수입니다.
* 12AssistStreakCount: 12회 연속 어시스트를 기록한 횟수입니다. 플레이어가 연속으로 12번 이상 어시스트한 경우 카운트됩니다.
* InfernalScalePickup: 특정 아이템 또는 강화 요소와 관련된 변수일 가능성이 있지만, 구체적인 기능은 보통 패치나 모드에 따라 달라집니다.
* acesBefore15Minutes: 게임 시작 후 15분 이내에 발생한 에이스 횟수입니다.
* alliedJungleMonsterKills: 플레이어가 아군 정글에서 처치한 몬스터의 수입니다.
* bountyGold: 플레이어가 현재 현상금으로 가지고 있는 골드 양입니다. 킬이나 연속 처치로 인해 쌓인 현상금을 의미합니다.
* buffsStolen: 상대팀 정글에서 훔친 버프의 수입니다. 예를 들어 블루 버프나 레드 버프 등을 포함합니다.
* completeSupportQuestInTime: 서포트 퀘스트를 지정된 시간 내에 완료했는지 여부를 나타냅니다. 서포트 전용 퀘스트 아이템과 관련됩니다.
* dancedWithRiftHerald: 리프트 헤럴드 근처에서 춤을 춘 횟수를 나타냅니다. 주로 유머나 이벤트적인 요소로 포함될 수 있는 변수입니다.
* doubleAces: 양 팀이 동시에 에이스를 기록한 횟수입니다.
* dragonTakedowns: 플레이어가 처치한 드래곤의 수입니다. 최종 타격을 기록한 경우 포함됩니다.
* elderDragonKillsWithOpposingSoul: 상대 팀이 드래곤 영혼을 가지고 있는 상태에서 처치한 장로 드래곤의 수입니다.
* elderDragonMultikills: 장로 드래곤과 관련된 멀티킬 횟수입니다.
* epicMonsterKillsNearEnemyJungler: 상대 정글러 근처에서 처치한 에픽 몬스터(바론, 드래곤, 리프트 헤럴드) 수입니다.
* epicMonsterKillsWithin30SecondsOfSpawn: 생성 후 30초 이내에 처치한 에픽 몬스터의 수입니다.
* epicMonsterSteals: 상대팀이 처치하려던 에픽 몬스터를 훔친 횟수입니다.
* epicMonsterStolenWithoutSmite: 스마이트 없이 에픽 몬스터를 스틸한 횟수입니다.
* fistBumpParticipation: 팀원과 협력하여 기록한 성과를 나타내는 변수입니다. (특정 이벤트와 연관될 수 있음)
* flawlessAces: 플로리스 에이스(사망자가 없는 에이스)를 기록한 횟수입니다.
* getTakedownsInAllLanesEarlyJungleAsLaner: 초반 정글 단계에서 모든 라인에서 처치에 기여한 횟수를 나타냅니다.
* hadOpenNexus: 넥서스가 개방된 상태에서 게임을 진행한 적이 있는지 여부를 나타냅니다.
* initialBuffCount: 게임 초반에 획득한 버프 수입니다. 블루 버프나 레드 버프와 같은 정글 버프를 초기 단계에서 얻은 횟수를 나타냅니다.
* initialCrabCount: 초기 단계에서 스컬 크랩(강가에 나타나는 중립 몬스터)을 처치한 횟수입니다.
* junglerTakedownsNearDamagedEpicMonster: 손상된 에픽 몬스터(드래곤, 바론 등) 근처에서 정글러가 처치에 기여한 횟수를 나타냅니다.
* challengeKillingSprees: 도전 과제와 연관된 연속 킬 횟수입니다.
* killsNearEnemyTurret: 적 타워 근처에서 적 챔피언을 처치한 횟수입니다.
* killsOnOtherLanesEarlyJungleAsLaner: 초반 정글 단계에서 다른 라인에서 획득한 킬 수입니다.
* killsOnRecentlyHealedByAramPack: ARAM 패키지(체력 회복 오브젝트)를 사용한 후 적을 처치한 횟수입니다.
* killsUnderOwnTurret: 아군 타워 아래에서 적을 처치한 횟수입니다.
* killsWithHelpFromEpicMonster: 에픽 몬스터(예: 바론, 드래곤)의 도움으로 적을 처치한 횟수입니다.
* knockEnemyIntoTeamAndKill: 적을 팀 쪽으로 밀어 넣은 후 처치한 횟수입니다.
* legendaryCount: 전설적인 킬(연속 킬 횟수) 수를 나타냅니다.
* lostAnInhibitor: 팀이 억제기를 잃었는지 여부를 나타냅니다.
* moreEnemyJungleThanOpponent: 상대보다 더 많은 적 정글 몬스터를 처치했는지 여부를 나타냅니다.
* mejaisFullStackInTime: 메자이의 영혼약자 아이템을 풀 스택으로 쌓은 시간입니다.
* multiKillOneSpell: 한 번의 스킬 사용으로 다중 킬(2명 이상)을 기록한 횟수입니다.
* multiTurretRiftHeraldCount: 리프트 헤럴드가 여러 타워를 파괴한 횟수입니다.
* multikills: 멀티킬(더블 킬, 트리플 킬 등)을 기록한 횟수입니다.
* multikillsAfterAggressiveFlash: 공격적인 플래시(점멸) 이후 기록한 멀티킬 횟수입니다.
* outerTurretExecutesBefore10Minutes: 게임 시작 후 10분 이내에 외곽 타워를 파괴한 횟수입니다.
* outnumberedNexusKill: 적이 수적으로 우세한 상황에서 넥서스를 파괴한 횟수입니다.
* perfectDragonSoulsTaken: 완벽한 드래곤 영혼 획득 횟수입니다.
* perfectGame: 무사고 승리(플레이어가 한 번도 죽지 않고 게임을 이긴 경우) 여부를 나타냅니다.
* pickKillWithAlly: 아군과 함께 적을 처치한 횟수입니다.
* playedChampSelectPosition: 챔피언 선택에서 플레이한 포지션을 나타냅니다.
* poroExplosions: 포로 폭발 횟수입니다. 주로 유머나 특정 이벤트에서 나타날 수 있습니다.
* quickCleanse: 빠르게 클렌즈(정화)를 사용한 횟수입니다.
* quickFirstTurret: 매우 빠르게 첫 번째 타워를 파괴한 횟수입니다.
* quickSoloKills: 빠르게 단독 킬을 기록한 횟수입니다.
* riftHeraldTakedowns: 리프트 헤럴드를 처치한 횟수입니다.
* saveAllyFromDeath: 아군을 죽음으로부터 구한 횟수입니다.
* scuttleCrabKills: 스컬 크랩(강가에 나타나는 중립 몬스터)을 처치한 횟수입니다.
* snowballsHit: 스킬인 스노우볼(눈덩이)을 맞춘 횟수입니다 (주로 ARAM 모드에서 사용됨).
* soloBaronKills: 혼자서 바론을 처치한 횟수입니다.
* survivedSingleDigitHpCount: 체력이 한 자릿수인 상태로 생존한 횟수입니다.
* takedownOnFirstTurret: 첫 번째 타워 근처에서 적을 처치한 횟수입니다.
* takedowns: 적을 처치하거나 어시스트한 총 횟수입니다. 일반적으로 킬과 어시스트를 포함합니다.
* takedownsAfterGainingLevelAdvantage: 레벨 우위를 얻은 후 적을 처치하거나 어시스트한 횟수입니다.
* takedownsBeforeJungleMinionSpawn: 정글 미니언이 생성되기 전에 적을 처치하거나 어시스트한 횟수입니다.
* takedownsFirstXMinutes: 게임 초반 특정 시간 내에 적을 처치하거나 어시스트한 횟수입니다.
* takedownsInAlcove: 알코브 위치에서 적을 처치하거나 어시스트한 횟수입니다. 알코브는 소환사의 협곡 맵의 상단과 하단에 위치한 구역입니다.
* tookLargeDamageSurvived: 큰 피해를 받고도 생존한 횟수입니다. 강력한 공격을 받고도 살아남은 경우를 기록합니다.
* challengeTurretTakedowns: 타워 파괴 도전에 성공한 횟수입니다. 여러 타워를 연속해서 파괴하거나 특별한 상황에서 타워를 파괴한 경우를 의미할 수 있습니다.
* turretsTakenWithRiftHerald: 리프트 헤럴드를 이용해 파괴한 타워의 수입니다.
* twentyMinionsIn3SecondsCount: 3초 이내에 20개의 미니언을 처치한 횟수입니다. 매우 빠른 속도로 미니언을 처치한 상황을 나타냅니다.
* twoWardsOneSweeperCount: 두 개의 와드와 하나의 시야 제거기를 동시에 사용한 횟수입니다.
* unseenRecalls: 적이 플레이어의 리콜(기지 귀환)을 시야에서 보지 못한 횟수입니다.
* voidMonsterKill: 공허 속성 몬스터를 처치한 횟수입니다. 바론 또는 특정 이벤트 몬스터가 포함될 수 있습니다.



## **써먹고 바로 삭제**
* eligibleForProgression: 게임에서 진행도 적립이 가능한지 여부.(정상적인 게임 참여 여부 False 날리기)
* gameEndedInEarlySurrender: 게임이 조기 항복으로 종료되었는지 여부.(True 날리기)
* teamEarlySurrendered: 내 팀이 조기 항복했는지 여부.(True 날리기)



In [4]:
# 삭제할 컬럼 목록
# 게임 외부 요소들(내부적으로 써먹을 수 없는) + 중복값들 + 극심한 데이터 불균형으로 인한 인사이트를 도출할 수 없는 + 인 게임 데이터이나 목적에 맞지 않는 변수들(이모지 사용, 춤 춘 횟수 등)

columns_to_drop = [
    'leaguePoints', 'championId', 'championTransform', 'gameEndedInSurrender', 'nexusKills', 'nexusTakedowns',
    'participantId', 'pentaKills', 'quadraKills', 'profileIcon', 'teamId', 'tripleKills', 'unrealKills',
    'baronKills', 'baronTakedowns', 'blastConeOppositeOpponentCount', '12AssistStreakCount', 'InfernalScalePickup',
    'acesBefore15Minutes', 'alliedJungleMonsterKills', 'bountyGold', 'buffsStolen', 'completeSupportQuestInTime',
    'dancedWithRiftHerald', 'doubleAces', 'dragonTakedowns', 'elderDragonKillsWithOpposingSoul', 'elderDragonMultikills',
    'epicMonsterKillsNearEnemyJungler', 'epicMonsterKillsWithin30SecondsOfSpawn', 'epicMonsterSteals',
    'epicMonsterStolenWithoutSmite', 'fistBumpParticipation', 'flawlessAces', 'getTakedownsInAllLanesEarlyJungleAsLaner',
    'hadOpenNexus', 'initialBuffCount', 'initialCrabCount', 'junglerTakedownsNearDamagedEpicMonster',
    'challengeKillingSprees', 'killsNearEnemyTurret', 'killsOnOtherLanesEarlyJungleAsLaner',
    'killsOnRecentlyHealedByAramPack', 'killsUnderOwnTurret', 'killsWithHelpFromEpicMonster',
    'knockEnemyIntoTeamAndKill', 'legendaryCount', 'lostAnInhibitor', 'moreEnemyJungleThanOpponent',
    'mejaisFullStackInTime', 'multiKillOneSpell', 'multiTurretRiftHeraldCount', 'multikills',
    'multikillsAfterAggressiveFlash', 'outerTurretExecutesBefore10Minutes', 'outnumberedNexusKill',
    'perfectDragonSoulsTaken', 'perfectGame', 'pickKillWithAlly', 'poroExplosions',
    'quickCleanse', 'quickFirstTurret', 'quickSoloKills', 'riftHeraldTakedowns', 'saveAllyFromDeath',
    'scuttleCrabKills', 'snowballsHit', 'soloBaronKills', 'survivedSingleDigitHpCount', 'takedownOnFirstTurret',
    'takedowns', 'takedownsAfterGainingLevelAdvantage', 'takedownsBeforeJungleMinionSpawn', 'takedownsFirstXMinutes',
    'takedownsInAlcove', 'tookLargeDamageSurvived', 'challengeTurretTakedowns', 'turretsTakenWithRiftHerald',
    'twentyMinionsIn3SecondsCount', 'twoWardsOneSweeperCount', 'unseenRecalls', 'voidMonsterKill'
]

data = data.drop(columns=columns_to_drop)

data.head()

tier rank  champExperience  champLevel championName  commandPings  \
0  BRONZE    I            14444          15        Poppy             2   
1  BRONZE    I             7904          11        Braum             2   
2  BRONZE    I            12153          14      Morgana             1   
3  BRONZE    I             8014          11       Thresh             3   
4  BRONZE    I            11402          13        Ziggs             1   

   consumablesPurchased  damageDealtToBuildings  damageDealtToObjectives  \
0                     4                    1291                     1396   
1                     6                       0                        0   
2                    10                    1515                     1922   
3                     5                    1420                     2703   
4                     6                     530                      951   

   damageDealtToTurrets  ...  teamBaronKills  teamDamagePercentage  \
0                  1291  ...               0              0.136533   
1                     0  ...               0              0.122879   
2                  1515  ...               0              0.174090   
3                  1420  ...               0              0.071151   
4                   530  ...               0              0.156749   

   teamElderDragonKills  teamRiftHeraldKills  turretPlatesTaken  \
0                     0                    0                  2   
1                     0                    1                  1   
2                     0                    0                  1   
3                     0                    1                  0   
4                     0                    0                  1   

   visionScoreAdvantageLaneOpponent  visionScorePerMinute  wardTakedowns  \
0                          0.326022              0.683416              1   
1                         -0.420499              1.147305              3   
2                         -0.135613              1.542737              3   
3                         -0.278238              1.656584              4   
4                         -0.552488              0.912949              2   

   wardTakedownsBefore20M  wardsGuarded  
0                       0             0  
1                       1             0  
2                       1             0  
3                       3             0  
4                       2             0  

[5 rows x 161 columns]

#### SWARM

* SWARM은 RANKED_SOLO_5x5와 다른 게임 모드 자체가 다르다.
* challenge 내부의 값들로 challenge 변수의 특성을 고려할 때, 한번에 기록된 듯 싶다.
* 이외에도 한 게임 내에서 이루어진 지표가 아닌, 누적 지표들이 있을 수 있기에 유의해야한다.

In [5]:
data.columns[data.columns.str.contains('SWARM', case=False)]

Index(['SWARM_DefeatAatrox', 'SWARM_DefeatBriar', 'SWARM_DefeatMiniBosses',
       'SWARM_EvolveWeapon', 'SWARM_Have3Passives', 'SWARM_KillEnemy',
       'SWARM_PickupGold', 'SWARM_ReachLevel50', 'SWARM_Survive15Min',
       'SWARM_WinWith5EvolvedWeapons'],
      dtype='object')

In [6]:
x = pd.DataFrame()
x['columns'] = data.columns

In [7]:
y = x.query("columns.str.contains('SWARM',case=False)")
y['columns']

104              SWARM_DefeatAatrox
105               SWARM_DefeatBriar
106          SWARM_DefeatMiniBosses
107              SWARM_EvolveWeapon
108             SWARM_Have3Passives
109                 SWARM_KillEnemy
110                SWARM_PickupGold
111              SWARM_ReachLevel50
112              SWARM_Survive15Min
113    SWARM_WinWith5EvolvedWeapons
Name: columns, dtype: object

In [8]:
data.shape

(178901, 161)

In [9]:
data.drop(columns=y['columns'],axis=1,inplace=True)

In [10]:
data.reset_index(drop=True,inplace=True)

In [11]:
data.shape

(178901, 151)

### 타겟 변수

* 브론즈1과 브론즈4의 뚜렷한 실력 차이가 존재하는가.

* 만약 존재한다면, 다이아1과 다이아4의 뚜렷한 실력 차이와 비교하여 어느 정도의 수준인가.

* 나머지 티어에서는 ?

In [13]:
low_dia = data.loc[(data['tier'] == 'DIAMOND') & (data['rank'] == 'I') | (data['rank'] == 'II')]
high_dia = data.loc[(data['tier'] == 'DIAMOND') & (data['rank'] == 'III') | (data['rank'] == 'IV')]

In [14]:
columns_to_compare = data.select_dtypes(include=['int64', 'float64']).columns.tolist()

In [ ]:
for col in columns_to_compare:
    plt.figure(figsize=(10, 6))
    plt.hist(high_dia[col], bins=20, alpha=0.5, label='High Diamond (I, II)', density=True)
    plt.hist(low_dia[col], bins=20, alpha=0.5, label='Low Diamond (III, IV)', density=True)
    plt.xlabel(col)
    plt.ylabel('Density')
    plt.title(f'Comparison of {col} between High and Low Diamond Tiers')
    plt.legend()
    plt.show()

Output hidden; open in https://colab.research.google.com to view.

> 결론적으로 DIAMOND의 세부 rank 별 차이는 분명 존재하나, 티어가 낮아질수록 뚜렷한 경향은 낮아질 것이다.

In [15]:
data = data.drop(columns='rank')

In [16]:
data.shape

(178901, 150)

In [17]:
data['tier'].value_counts()

tier
DIAMOND     37495
EMERALD     29535
SILVER      28860
GOLD        28097
PLATINUM    27857
BRONZE      27057
Name: count, dtype: int64

* ability 요소들의 total값과 abilityUses의 합이 다르다.
* abilityUses : 특정 중요 상황에서만 카운트했다-라는데 직관적인 이해불가로 drop


In [ ]:
totalabil =data['spell1Casts']+data['spell2Casts']+data['spell3Casts']+data['spell4Casts']+data['summoner1Casts']+data['summoner2Casts']
totalabil.max()

1911

In [ ]:
data['abilityUses'].max()

1879

In [18]:
data = data.drop(columns='abilityUses')
data.head()

tier  champExperience  champLevel championName  commandPings  \
0  BRONZE            14444          15        Poppy             2   
1  BRONZE             7904          11        Braum             2   
2  BRONZE            12153          14      Morgana             1   
3  BRONZE             8014          11       Thresh             3   
4  BRONZE            11402          13        Ziggs             1   

   consumablesPurchased  damageDealtToBuildings  damageDealtToObjectives  \
0                     4                    1291                     1396   
1                     6                       0                        0   
2                    10                    1515                     1922   
3                     5                    1420                     2703   
4                     6                     530                      951   

   damageDealtToTurrets  damageSelfMitigated  ...  teamBaronKills  \
0                  1291                37543  ...               0   
1                     0                19919  ...               0   
2                  1515                14613  ...               0   
3                  1420                13896  ...               0   
4                   530                12118  ...               0   

   teamDamagePercentage  teamElderDragonKills  teamRiftHeraldKills  \
0              0.136533                     0                    0   
1              0.122879                     0                    1   
2              0.174090                     0                    0   
3              0.071151                     0                    1   
4              0.156749                     0                    0   

   turretPlatesTaken  visionScoreAdvantageLaneOpponent  visionScorePerMinute  \
0                  2                          0.326022              0.683416   
1                  1                         -0.420499              1.147305   
2                  1                         -0.135613              1.542737   
3                  0                         -0.278238              1.656584   
4                  1                         -0.552488              0.912949   

   wardTakedowns  wardTakedownsBefore20M  wardsGuarded  
0              1                       0             0  
1              3                       1             0  
2              3                       1             0  
3              4                       3             0  
4              2                       2             0  

[5 rows x 149 columns]

In [19]:
# 정상적인 게임 참여 여부(소환사 경험치 적립 가능 여부)
data['eligibleForProgression'].value_counts()

eligibleForProgression
True     175793
False      3108
Name: count, dtype: int64

In [20]:
data = data[data['eligibleForProgression'] != False].reset_index(drop=True)
data

tier  champExperience  champLevel championName  commandPings  \
0        BRONZE            14444          15        Poppy             2   
1        BRONZE             7904          11        Braum             2   
2        BRONZE            12153          14      Morgana             1   
3        BRONZE             8014          11       Thresh             3   
4        BRONZE            11402          13        Ziggs             1   
...         ...              ...         ...          ...           ...   
175788  DIAMOND             5125           9        Poppy             7   
175789  DIAMOND             9991          13         Zyra            11   
175790  DIAMOND             7181          10       Soraka             3   
175791  DIAMOND            14246          15        Leona            53   
175792  DIAMOND            13284          15       Aurora             0   

        consumablesPurchased  damageDealtToBuildings  damageDealtToObjectives  \
0                          4                    1291                     1396   
1                          6                       0                        0   
2                         10                    1515                     1922   
3                          5                    1420                     2703   
4                          6                     530                      951   
...                      ...                     ...                      ...   
175788                     5                     229                     1622   
175789                     7                    1534                     8994   
175790                     5                      44                      319   
175791                    15                     479                      941   
175792                     0                     684                      684   

        damageDealtToTurrets  damageSelfMitigated  ...  teamBaronKills  \
0                       1291                37543  ...               0   
1                          0                19919  ...               0   
2                       1515                14613  ...               0   
3                       1420                13896  ...               0   
4                        530                12118  ...               0   
...                      ...                  ...  ...             ...   
175788                   229                 6695  ...               0   
175789                  1534                 7639  ...               1   
175790                    44                 6344  ...               0   
175791                   479                65088  ...               2   
175792                   684                12809  ...               0   

        teamDamagePercentage  teamElderDragonKills  teamRiftHeraldKills  \
0                   0.136533                     0                    0   
1                   0.122879                     0                    1   
2                   0.174090                     0                    0   
3                   0.071151                     0                    1   
4                   0.156749                     0                    0   
...                      ...                   ...                  ...   
175788              0.093600                     0                    0   
175789              0.156914                     0                    0   
175790              0.033654                     0                    1   
175791              0.144250                     0                    1   
175792              0.293750                     0                    0   

        turretPlatesTaken  visionScoreAdvantageLaneOpponent  \
0                       2                          0.326022   
1                       1                         -0.420499   
2                       1                         -0.135613   
3                       0                         -0.278238   
4                       1       

In [21]:
# 게임이 조기 항복으로 종료되었는지 여부
# 위 eligibleForProgression 변수에서 다 걸러졌다.
data['gameEndedInEarlySurrender'].value_counts()

gameEndedInEarlySurrender
False    175793
Name: count, dtype: int64

In [ ]:
# 내 팀이 조기 항복했는지 여부
data['teamEarlySurrendered'].value_counts()

teamEarlySurrendered
False    175793
Name: count, dtype: int64

In [22]:
data = data.drop(columns=['eligibleForProgression','gameEndedInEarlySurrender','teamEarlySurrendered'])
data

tier  champExperience  champLevel championName  commandPings  \
0        BRONZE            14444          15        Poppy             2   
1        BRONZE             7904          11        Braum             2   
2        BRONZE            12153          14      Morgana             1   
3        BRONZE             8014          11       Thresh             3   
4        BRONZE            11402          13        Ziggs             1   
...         ...              ...         ...          ...           ...   
175788  DIAMOND             5125           9        Poppy             7   
175789  DIAMOND             9991          13         Zyra            11   
175790  DIAMOND             7181          10       Soraka             3   
175791  DIAMOND            14246          15        Leona            53   
175792  DIAMOND            13284          15       Aurora             0   

        consumablesPurchased  damageDealtToBuildings  damageDealtToObjectives  \
0                          4                    1291                     1396   
1                          6                       0                        0   
2                         10                    1515                     1922   
3                          5                    1420                     2703   
4                          6                     530                      951   
...                      ...                     ...                      ...   
175788                     5                     229                     1622   
175789                     7                    1534                     8994   
175790                     5                      44                      319   
175791                    15                     479                      941   
175792                     0                     684                      684   

        damageDealtToTurrets  damageSelfMitigated  ...  teamBaronKills  \
0                       1291                37543  ...               0   
1                          0                19919  ...               0   
2                       1515                14613  ...               0   
3                       1420                13896  ...               0   
4                        530                12118  ...               0   
...                      ...                  ...  ...             ...   
175788                   229                 6695  ...               0   
175789                  1534                 7639  ...               1   
175790                    44                 6344  ...               0   
175791                   479                65088  ...               2   
175792                   684                12809  ...               0   

        teamDamagePercentage  teamElderDragonKills  teamRiftHeraldKills  \
0                   0.136533                     0                    0   
1                   0.122879                     0                    1   
2                   0.174090                     0                    0   
3                   0.071151                     0                    1   
4                   0.156749                     0                    0   
...                      ...                   ...                  ...   
175788              0.093600                     0                    0   
175789              0.156914                     0                    0   
175790              0.033654                     0                    1   
175791              0.144250                     0                    1   
175792              0.293750                     0                    0   

        turretPlatesTaken  visionScoreAdvantageLaneOpponent  \
0                       2                          0.326022   
1                       1                         -0.420499   
2                       1                         -0.135613   
3                       0                         -0.278238   
4                       1       

* gameLength: 게임의 총 길이(초 단위)입니다. 한 판이 끝날 때까지의 총 시간을 나타냅니다.(모든 소환사 동일)
* timePlayed: 플레이한 총 시간.(접속 끊김,도주 등 실제 플레이 시간 기록)

In [ ]:
data['timePlayed']

0         1831
1         1645
2         1971
3         1548
4         1751
          ... 
175788     970
175789    1581
175790    1339
175791    2485
175792     933
Name: timePlayed, Length: 175793, dtype: int64

In [ ]:
data['gameLength']

0         1831.725851
1         1645.296915
2         1971.431968
3         1548.736697
4         1751.642870
             ...     
175788     970.856135
175789    1582.120136
175790    1339.895719
175791    2485.801991
175792     933.945902
Name: gameLength, Length: 175793, dtype: float64

In [ ]:
data['gameLength'] = data['gameLength'].astype(int)

In [ ]:
x = pd.DataFrame()
x['time'] = data['gameLength']-data['timePlayed']
# 한 차례 걸러진 데이터에서는 실질적인 게임 시간과 플레이 시간의 차이는 8초 이내. 그냥 timePlayed 변수 drop
x[x['time']>=8]

Empty DataFrame
Columns: [time]
Index: []

In [23]:
data = data.drop(columns='timePlayed')
data

tier  champExperience  champLevel championName  commandPings  \
0        BRONZE            14444          15        Poppy             2   
1        BRONZE             7904          11        Braum             2   
2        BRONZE            12153          14      Morgana             1   
3        BRONZE             8014          11       Thresh             3   
4        BRONZE            11402          13        Ziggs             1   
...         ...              ...         ...          ...           ...   
175788  DIAMOND             5125           9        Poppy             7   
175789  DIAMOND             9991          13         Zyra            11   
175790  DIAMOND             7181          10       Soraka             3   
175791  DIAMOND            14246          15        Leona            53   
175792  DIAMOND            13284          15       Aurora             0   

        consumablesPurchased  damageDealtToBuildings  damageDealtToObjectives  \
0                          4                    1291                     1396   
1                          6                       0                        0   
2                         10                    1515                     1922   
3                          5                    1420                     2703   
4                          6                     530                      951   
...                      ...                     ...                      ...   
175788                     5                     229                     1622   
175789                     7                    1534                     8994   
175790                     5                      44                      319   
175791                    15                     479                      941   
175792                     0                     684                      684   

        damageDealtToTurrets  damageSelfMitigated  ...  teamBaronKills  \
0                       1291                37543  ...               0   
1                          0                19919  ...               0   
2                       1515                14613  ...               0   
3                       1420                13896  ...               0   
4                        530                12118  ...               0   
...                      ...                  ...  ...             ...   
175788                   229                 6695  ...               0   
175789                  1534                 7639  ...               1   
175790                    44                 6344  ...               0   
175791                   479                65088  ...               2   
175792                   684                12809  ...               0   

        teamDamagePercentage  teamElderDragonKills  teamRiftHeraldKills  \
0                   0.136533                     0                    0   
1                   0.122879                     0                    1   
2                   0.174090                     0                    0   
3                   0.071151                     0                    1   
4                   0.156749                     0                    0   
...                      ...                   ...                  ...   
175788              0.093600                     0                    0   
175789              0.156914                     0                    0   
175790              0.033654                     0                    1   
175791              0.144250                     0                    1   
175792              0.293750                     0                    0   

        turretPlatesTaken  visionScoreAdvantageLaneOpponent  \
0                       2                          0.326022   
1                       1                         -0.420499   
2                       1                         -0.135613   
3                       0                         -0.278238   
4                       1       

In [24]:
data.shape

(175793, 145)

### 중복값 제거

In [ ]:
data[data.duplicated()]

tier  champExperience  champLevel championName  commandPings  \
618      BRONZE            14891          16         Yone             1   
619      BRONZE            13718          15         Kayn             7   
620      BRONZE             4536           8        Swain             0   
621      BRONZE            11146          13         Yone             5   
622      BRONZE             8439          11       Irelia             6   
...         ...              ...         ...          ...           ...   
173639  DIAMOND            14304          15         Jinx            18   
173640  DIAMOND            15832          16           Vi            19   
173641  DIAMOND             6675          10        Braum             9   
173642  DIAMOND            20266          18        Shaco             0   
173643  DIAMOND            11564          14           Vi             5   

        consumablesPurchased  damageDealtToBuildings  damageDealtToObjectives  \
618                        0                     671                      671   
619                        0                    3573                    38628   
620                        6                     585                      585   
621                        1                     533                     1824   
622                        1                    6307                     6307   
...                      ...                     ...                      ...   
173639                     2                    5738                    18053   
173640                     6                    3629                     6393   
173641                     4                     795                     1013   
173642                     0                     108                      108   
173643                     1                    7887                     7887   

        damageDealtToTurrets  damageSelfMitigated  ...  teamBaronKills  \
618                      671                28943  ...               0   
619                     3573                26725  ...               0   
620                      585                 8519  ...               0   
621                      533                12438  ...               0   
622                     6307                 8037  ...               0   
...                      ...                  ...  ...             ...   
173639                  5738                17062  ...               0   
173640                  3629                41374  ...               1   
173641                   795                 7621  ...               0   
173642                   108                16143  ...               0   
173643                  7887                22365  ...               0   

        teamDamagePercentage  teamElderDragonKills  teamRiftHeraldKills  \
618                 0.182789                     0                    0   
619                 0.182795                     0                    1   
620                 0.120263                     0                    0   
621                 0.227054                     0                    0   
622                 0.238293                     0                    0   
...                      ...                   ...                  ...   
173639              0.300168                     0                    0   
173640              0.123624                     0                    1   
173641              0.235365                     0                    0   
173642              0.135346                     0                    0   
173643              0.266833                     0                    0   

        turretPlatesTaken  visionScoreAdvantageLaneOpponent  \
618                     0                               NaN   
619                     0                         -0.454944   
620                     2                          0.035816   
621                     1                         -0.374505   
622                     6       

In [ ]:
data[(data['championName']=='Yone')&(data['champExperience']==14891)]

tier  champExperience  champLevel championName  commandPings  \
578  BRONZE            14891          16         Yone             1   
618  BRONZE            14891          16         Yone             1   

     consumablesPurchased  damageDealtToBuildings  damageDealtToObjectives  \
578                     0                     671                      671   
618                     0                     671                      671   

     damageDealtToTurrets  damageSelfMitigated  ...  teamBaronKills  \
578                   671                28943  ...               0   
618                   671                28943  ...               0   

     teamDamagePercentage  teamElderDragonKills  teamRiftHeraldKills  \
578              0.182789                     0                    0   
618              0.182789                     0                    0   

     turretPlatesTaken  visionScoreAdvantageLaneOpponent  \
578                  0                               NaN   
618                  0                               NaN   

     visionScorePerMinute  wardTakedowns  wardTakedownsBefore20M  wardsGuarded  
578                   0.0              0                       0             0  
618                   0.0              0                       0             0  

[2 rows x 145 columns]

In [ ]:
data[(data['championName']=='Kayn')&(data['champExperience']==13718)]

tier  champExperience  champLevel championName  commandPings  \
579  BRONZE            13718          15         Kayn             7   
619  BRONZE            13718          15         Kayn             7   

     consumablesPurchased  damageDealtToBuildings  damageDealtToObjectives  \
579                     0                    3573                    38628   
619                     0                    3573                    38628   

     damageDealtToTurrets  damageSelfMitigated  ...  teamBaronKills  \
579                  3573                26725  ...               0   
619                  3573                26725  ...               0   

     teamDamagePercentage  teamElderDragonKills  teamRiftHeraldKills  \
579              0.182795                     0                    1   
619              0.182795                     0                    1   

     turretPlatesTaken  visionScoreAdvantageLaneOpponent  \
579                  0                         -0.454944   
619                  0                         -0.454944   

     visionScorePerMinute  wardTakedowns  wardTakedownsBefore20M  wardsGuarded  
579              0.446745              0                       0             0  
619              0.446745              0                       0             0  

[2 rows x 145 columns]

In [25]:
data.drop_duplicates(inplace=True)

In [26]:
data.reset_index(drop=True, inplace=True)

In [27]:
data.shape

(172824, 145)

In [28]:
data[data.duplicated()]

Empty DataFrame
Columns: [tier, champExperience, champLevel, championName, commandPings, consumablesPurchased, damageDealtToBuildings, damageDealtToObjectives, damageDealtToTurrets, damageSelfMitigated, dangerPings, deaths, detectorWardsPlaced, doubleKills, dragonKills, enemyMissingPings, enemyVisionPings, firstBloodAssist, firstBloodKill, firstTowerAssist, firstTowerKill, getBackPings, goldEarned, goldSpent, holdPings, individualPosition, inhibitorKills, inhibitorTakedowns, inhibitorsLost, item0, item1, item2, item3, item4, item5, item6, itemsPurchased, killingSprees, kills, lane, largestCriticalStrike, largestKillingSpree, largestMultiKill, longestTimeSpentLiving, magicDamageDealt, magicDamageDealtToChampions, magicDamageTaken, needVisionPings, neutralMinionsKilled, nexusLost, objectivesStolen, objectivesStolenAssists, onMyWayPings, physicalDamageDealt, physicalDamageDealtToChampions, physicalDamageTaken, pushPings, role, sightWardsBoughtInGame, spell1Casts, spell2Casts, spell3Casts, spell4Casts, summoner1Casts, summoner1Id, summoner2Casts, summoner2Id, summonerLevel, teamPosition, timeCCingOthers, totalAllyJungleMinionsKilled, totalDamageDealt, totalDamageDealtToChampions, totalDamageShieldedOnTeammates, totalDamageTaken, totalEnemyJungleMinionsKilled, totalHeal, totalHealsOnTeammates, totalMinionsKilled, totalTimeCCDealt, totalTimeSpentDead, totalUnitsHealed, trueDamageDealt, trueDamageDealtToChampions, trueDamageTaken, turretKills, turretTakedowns, turretsLost, visionClearedPings, visionScore, visionWardsBoughtInGame, wardsKilled, wardsPlaced, win, allInPings, assistMePings, assists, basicPings, bountyLevel, controlWardsPlaced, ...]
Index: []

[0 rows x 145 columns]

### **라인과 포지션에 대한 고찰**

* teamPosition : 팀 내에서 사전 설정된 공식적인 포지션(매칭되었을 시 결정되는)

* role : 특정 라인 역할을 설명

    * 원딜 **DUO_CARRY**
    * 서포터 **DUO_SUPPORT**
    * 정글러 **NONE**

* lane : 플레이어가 가장 많이 머문 라인.
    * TOP, JUNGLE, MIDDLE, BOTTOM, NONE으로 구분
    * 특정 상황에서는 여러 라인에서 활동한 경우 NONE

* individualPosition:게임 내에서 플레이어가 실제로 수행한 포지션
    * teamPosition과 달리, 게임 도중의 실제 행동을 기반으로 포지션이 기록됩니다.

* playedChampSelectPosition : 챔피언 선택 단계에서 플레이어가 선택한 포지션
    * 매칭 시에 설정한 포지션(탑, 미드 등)
    * 이 변수는 게임이 시작되기 전에 선택한 포지션

In [ ]:
data['individualPosition'].value_counts()

individualPosition
JUNGLE     30365
TOP        29741
MIDDLE     29724
BOTTOM     29717
UTILITY    27482
Invalid    25795
Name: count, dtype: int64

In [ ]:
# Invalid : 25802 버리기 아까우니까 추정을 좀 해보자.

In [ ]:
data['lane'].value_counts()

lane
BOTTOM    44455
NONE      40696
JUNGLE    30875
MIDDLE    30403
TOP       26395
Name: count, dtype: int64

In [ ]:
data['role'].value_counts()

role
SUPPORT    63501
SOLO       40831
NONE       30961
CARRY      22761
DUO        14770
Name: count, dtype: int64

In [ ]:
data['teamPosition'].value_counts()

teamPosition
JUNGLE     29969
BOTTOM     29763
MIDDLE     29762
TOP        29758
UTILITY    27777
Name: count, dtype: int64

In [ ]:
data.loc[(data['individualPosition'] == 'Invalid') & (data['lane'] == 'TOP')]

tier  champExperience  champLevel championName  commandPings  \
47       BRONZE            28329          18      Chogath             0   
77       BRONZE            22161          18       Kennen             4   
79       BRONZE            24520          18        Garen             0   
92       BRONZE            26206          18     Katarina             6   
144      BRONZE            26316          18        Brand             1   
...         ...              ...         ...          ...           ...   
172569  DIAMOND            18980          18       Ezreal             3   
172575  DIAMOND            21507          18          Zoe             2   
172614  DIAMOND            21022          18       Kennen             4   
172615  DIAMOND            22951          18         Zeri             1   
172762  DIAMOND            26602          18          Zoe             4   

        consumablesPurchased  damageDealtToBuildings  damageDealtToObjectives  \
47                         0                    2644                     2644   
77                         0                     306                      306   
79                         2                       0                        0   
92                         3                     115                      115   
144                        4                     898                      898   
...                      ...                     ...                      ...   
172569                     1                    6832                     6832   
172575                     2                    2029                     2029   
172614                     0                    2889                     2889   
172615                     0                     947                      947   
172762                     2                    3099                     3099   

        damageDealtToTurrets  damageSelfMitigated  ...  teamBaronKills  \
47                      2644                69152  ...               0   
77                       306                17152  ...               0   
79                         0                96136  ...               0   
92                       115                41283  ...               0   
144                      898                17134  ...               0   
...                      ...                  ...  ...             ...   
172569                  6832                12093  ...               0   
172575                  2029                14474  ...               0   
172614                  2889                13076  ...               0   
172615                   947                14920  ...               0   
172762                  3099                21048  ...               0   

        teamDamagePercentage  teamElderDragonKills  teamRiftHeraldKills  \
47                  0.189835                     0                    0   
77                  0.206457                     0                    0   
79                  0.166602                     0                    0   
92                  0.212350                     0                    0   
144                 0.253733                     0                    0   
...                      ...                   ...                  ...   
172569              0.251840                     0                    0   
172575              0.205962                     0                    0   
172614              0.283887                     0                    0   
172615              0.394388                     0                    0   
172762              0.362350                     0                    0   

        turretPlatesTaken  visionScoreAdvantageLaneOpponent  \
47                      0                               NaN   
77                      0                               NaN   
79                      0                               NaN   
92                      0                               NaN   
144                     0       

In [ ]:
# 원딜 캐릭으로 탑을 간 멋쟁이들이 상당히 보인다.
data['championName'].loc[(data['individualPosition'] == 'Invalid') & (data['lane'] == 'TOP')].value_counts()

championName
Jhin                 67
Ezreal               64
Zed                  63
Zeri                 63
Yone                 62
Varus                60
Vayne                59
Yasuo                59
Caitlyn              59
Irelia               58
LeeSin               57
Smolder              57
Lucian               55
Xayah                54
Aatrox               54
Ahri                 52
Ashe                 51
Jinx                 50
Samira               49
Akali                49
Veigar               49
Graves               48
MissFortune          48
Jayce                46
Sylas                46
Kaisa                45
Leblanc              44
Brand                44
Sivir                43
Aphelios             43
Viego                43
Morgana              42
TwistedFate          42
KogMaw               42
Chogath              42
Katarina             42
Vladimir             41
Senna                40
Swain                40
Garen                40
Xerath               40
Nidalee              40
Karthus              39
Sett                 37
Pantheon             36
DrMundo              36
Gangplank            36
Maokai               35
Vex                  35
Ekko                 33
Twitch               33
Kayle                33
Hwei                 33
Tristana             33
Viktor               32
Lux                  32
Shaco                32
Azir                 32
Kennen               32
Darius               32
Lillia               31
Gragas               30
Syndra               30
AurelionSol          29
Kalista              29
Orianna              28
Nasus                28
Corki                28
Singed               28
Cassiopeia           27
Mordekaiser          27
Taliyah              27
FiddleSticks         27
Teemo                27
Talon                27
Gnar                 27
Diana                27
Neeko                26
Velkoz               26
Rumble               26
Riven                25
Khazix               25
Ziggs                25
Sion                 25
Illaoi               25
Amumu                24
Draven               24
Seraphine            24
Malphite             24
Aurora               24
MonkeyKing           24
Ryze                 24
Kassadin             24
KSante               24
Jax                  23
Volibear             22
Nilah                22
Nunu                 21
Fizz                 21
Zoe                  21
Fiora                20
Galio                20
Vi                   20
Shyvana              19
Karma                19
Zyra                 18
Kindred              18
MasterYi             17
Thresh               17
Gwen                 17
Anivia               17
Ornn                 17
Sejuani              16
Alistar              16
Lissandra            16
JarvanIV             16
Akshan               16
Qiyana               16
TahmKench            15
Poppy                15
XinZhao              15
Blitzcrank           15
Udyr                 15
Nautilus             15
Zac                  14
Rengar               14
Tryndamere           14
Renekton             14
Naafiri              13
Camille              13
Briar                13
Heimerdinger         13
Zilean               12
Kayn                 12
Malzahar             12
Annie                12
Skarner              11
Olaf                 10
Rammus                9
Rell                  9
Urgot                 9
Yorick                9
Hecarim               9
Bard                  9
Lulu                  9
Janna                 8
Quinn                 8
Yuumi                 8
Shen                  8
Warwick               7
Sona                  7
Braum                 6
Ivern                 6
Pyke                  6
Trundle               6
Elise                 6
Leona                 6
Nami                  6
Soraka                5
Kled                  5
RekSai                5
Milio                 4
Strawberry_Riven      4
Belveth               4
Rakan                 4
Renata                4
Noc

In [ ]:
data.loc[(data['individualPosition'] == 'Invalid') & (data['lane'] == 'MIDDLE')]

tier  champExperience  champLevel  championName  commandPings  \
89       BRONZE            35599          18         Braum             7   
91       BRONZE            21107          18       Leblanc             8   
138      BRONZE            28871          18      Malphite             5   
145      BRONZE            19627          18  Heimerdinger             3   
307      BRONZE            30537          18      Vladimir             4   
...         ...              ...         ...           ...           ...   
172504  DIAMOND            27076          18          Rell             3   
172566  DIAMOND            15000          16         Fiora             0   
172599  DIAMOND            32172          18        Ezreal             2   
172685  DIAMOND            18897          18       Leblanc             2   
172687  DIAMOND            19763          18        KogMaw             0   

        consumablesPurchased  damageDealtToBuildings  damageDealtToObjectives  \
89                         2                     636                      636   
91                         1                    1181                     1181   
138                        0                       0                        0   
145                        2                    1022                     1022   
307                        2                    1653                     1653   
...                      ...                     ...                      ...   
172504                     1                     444                      444   
172566                     0                       0                     1373   
172599                     3                    8917                     8917   
172685                     1                     432                      432   
172687                     0                    3758                     3758   

        damageDealtToTurrets  damageSelfMitigated  ...  teamBaronKills  \
89                       636               138497  ...               0   
91                      1181                23155  ...               0   
138                        0                34414  ...               0   
145                     1022                 6864  ...               0   
307                     1653                28678  ...               0   
...                      ...                  ...  ...             ...   
172504                   444               125081  ...               0   
172566                     0                45481  ...               0   
172599                  8917                47331  ...               0   
172685                   432                18808  ...               0   
172687                  3758                22925  ...               0   

        teamDamagePercentage  teamElderDragonKills  teamRiftHeraldKills  \
89                  0.133216                     0                    0   
91                  0.220989                     0                    0   
138                 0.249897                     0                    0   
145                 0.297207                     0                    0   
307                 0.219599                     0                    0   
...                      ...                   ...                  ...   
172504              0.091996                     0                    0   
172566              0.090441                     0                    0   
172599              0.359097                     0                    0   
172685              0.196986                     0                    0   
172687              0.306505                     0                    0   

        turretPlatesTaken  visionScoreAdvantageLaneOpponent  \
89                      0                               NaN   
91                      0                               NaN   
138                     0                               NaN   
145                     0                               NaN   
307                 

In [ ]:
data['championName'].loc[(data['individualPosition'] == 'Invalid') & (data['lane'] == 'MIDDLE')].value_counts()

championName
Ezreal                  72
Caitlyn                 58
Senna                   55
Varus                   55
Lux                     55
                        ..
Naafiri                  1
Strawberry_Seraphine     1
Strawberry_Aurora        1
Kled                     1
Strawberry_Xayah         1
Name: count, Length: 176, dtype: int64

In [ ]:
data.loc[(data['individualPosition'] == 'Invalid') & (data['lane'] == 'JUNGLE')]

tier  champExperience  champLevel championName  commandPings  \
85       BRONZE            22380          18          Jax            25   
150      BRONZE            20082          18        Annie             0   
589      BRONZE            25904          18          Zed             1   
748      BRONZE            23653          18       Xerath             2   
855      BRONZE            23131          18      Warwick             8   
...         ...              ...         ...          ...           ...   
171637  DIAMOND            14000          15         Sion             5   
171991  DIAMOND            22963          18          Zac             0   
172233  DIAMOND            19795          18     Vladimir             0   
172249  DIAMOND            31168          18       Rengar             3   
172251  DIAMOND            17521          17      Orianna             0   

        consumablesPurchased  damageDealtToBuildings  damageDealtToObjectives  \
85                         2                      67                       67   
150                        0                     688                      688   
589                        0                     848                      848   
748                        0                    1992                     1992   
855                        0                    2343                     2343   
...                      ...                     ...                      ...   
171637                     4                       0                     1035   
171991                     0                     513                      513   
172233                     2                      86                       86   
172249                     3                      32                       32   
172251                     3                     184                      184   

        damageDealtToTurrets  damageSelfMitigated  ...  teamBaronKills  \
85                        67                66893  ...               0   
150                      688                 8916  ...               0   
589                      848                32462  ...               0   
748                     1992                 7803  ...               0   
855                     2343               110868  ...               0   
...                      ...                  ...  ...             ...   
171637                     0               118962  ...               0   
171991                   513               134549  ...               0   
172233                    86                17626  ...               0   
172249                    32                26265  ...               0   
172251                   184                18521  ...               0   

        teamDamagePercentage  teamElderDragonKills  teamRiftHeraldKills  \
85                  0.174467                     0                    0   
150                 0.211842                     0                    0   
589                 0.254234                     0                    0   
748                 0.393770                     0                    0   
855                 0.150119                     0                    0   
...                      ...                   ...                  ...   
171637              0.157791                     0                    0   
171991              0.133928                     0                    0   
172233              0.145092                     0                    0   
172249              0.225047                     0                    0   
172251              0.122026                     0                    0   

        turretPlatesTaken  visionScoreAdvantageLaneOpponent  \
85                      0                               NaN   
150                     0                               NaN   
589                     0                               NaN   
748                     0                               NaN   
855                     0       

In [ ]:
data['championName'].loc[(data['individualPosition'] == 'Invalid') & (data['lane'] == 'JUNGLE')].value_counts()
# 몇몇은 챔피언만 봐도 정상적인 게임을 했으리란 기대가 들지 않는다. 아니면, 특정 메타에 유독 좋은 빌드를 따라간 것.

championName
Lucian       21
LeeSin       20
Sett         19
Sylas        18
Jhin         18
             ..
Riven         1
Nilah         1
Karthus       1
Lissandra     1
Nocturne      1
Name: count, Length: 167, dtype: int64

In [ ]:
data.loc[(data['individualPosition'] == 'Invalid') & (data['lane'] == 'BOTTOM')]
# 여기 역시 마찬가지

tier  champExperience  champLevel      championName  commandPings  \
591      BRONZE            29993          18              Fizz             6   
604      BRONZE            27821          18             Kaisa             0   
1493     BRONZE            36814          18             Teemo             0   
2321     BRONZE            26315          18             Janna             1   
2815     BRONZE            26008          18            Aatrox             2   
...         ...              ...         ...               ...           ...   
171592  DIAMOND            32145          18         Gangplank             3   
171627  DIAMOND            17000          18           Chogath             1   
171836  DIAMOND            25632          18             Viego             3   
172213  DIAMOND             6476          42  Strawberry_Riven             0   
172234  DIAMOND            29908          18              Fizz             1   

        consumablesPurchased  damageDealtToBuildings  damageDealtToObjectives  \
591                        0                    1844                     1844   
604                        0                    1331                     1331   
1493                       2                    2331                     2331   
2321                       0                     719                      719   
2815                       1                     364                      364   
...                      ...                     ...                      ...   
171592                     9                     823                      823   
171627                     8                       0                        0   
171836                     2                     727                      727   
172213                     0                       0                        0   
172234                     2                    1179                     1179   

        damageDealtToTurrets  damageSelfMitigated  ...  teamBaronKills  \
591                     1844                34287  ...               0   
604                     1331                17874  ...               0   
1493                    2331                18849  ...               0   
2321                     719                16083  ...               0   
2815                     364                70360  ...               0   
...                      ...                  ...  ...             ...   
171592                   823                47187  ...               0   
171627                     0               144251  ...               0   
171836                   727                67058  ...               0   
172213                     0                58924  ...               0   
172234                  1179                35280  ...               0   

        teamDamagePercentage  teamElderDragonKills  teamRiftHeraldKills  \
591                 0.238224                     0                    0   
604                 0.273859                     0                    0   
1493                0.164226                     0                    0   
2321                0.082419                     0                    0   
2815                0.131239                     0                    0   
...                      ...                   ...                  ...   
171592              0.357042                     0                    0   
171627              0.125080                     0                    0   
171836              0.223046                     0                    0   
172213              0.177778                     0                    0   
172234              0.175239                     0                    0   

        turretPlatesTaken  visionScoreAdvantageLaneOpponent  \
591                     0                               NaN   
604                     0                               NaN   
1493                    0                               NaN   
2321                    0          

In [ ]:
data['championName'].loc[(data['individualPosition'] == 'Invalid') & (data['lane'] == 'BOTTOM')].value_counts()
# 사실 제일 박살난 곳은 여기다.

championName
Kaisa               26
Jhin                20
Leblanc             17
LeeSin              17
Sett                15
                    ..
KogMaw               1
Corki                1
Gnar                 1
Yasuo                1
Strawberry_Riven     1
Name: count, Length: 164, dtype: int64

In [29]:
# 포기
# 배정받은대로 활동한 그나마 정상적인 게임들만 갖고가도록 하자.

data = data[data['individualPosition'] == data['teamPosition']].reset_index(drop=True)
data.head()

tier  champExperience  champLevel championName  commandPings  \
0  BRONZE            14444          15        Poppy             2   
1  BRONZE             7904          11        Braum             2   
2  BRONZE            12153          14      Morgana             1   
3  BRONZE             8014          11       Thresh             3   
4  BRONZE            11402          13        Ziggs             1   

   consumablesPurchased  damageDealtToBuildings  damageDealtToObjectives  \
0                     4                    1291                     1396   
1                     6                       0                        0   
2                    10                    1515                     1922   
3                     5                    1420                     2703   
4                     6                     530                      951   

   damageDealtToTurrets  damageSelfMitigated  ...  teamBaronKills  \
0                  1291                37543  ...               0   
1                     0                19919  ...               0   
2                  1515                14613  ...               0   
3                  1420                13896  ...               0   
4                   530                12118  ...               0   

   teamDamagePercentage  teamElderDragonKills  teamRiftHeraldKills  \
0              0.136533                     0                    0   
1              0.122879                     0                    1   
2              0.174090                     0                    0   
3              0.071151                     0                    1   
4              0.156749                     0                    0   

   turretPlatesTaken  visionScoreAdvantageLaneOpponent  visionScorePerMinute  \
0                  2                          0.326022              0.683416   
1                  1                         -0.420499              1.147305   
2                  1                         -0.135613              1.542737   
3                  0                         -0.278238              1.656584   
4                  1                         -0.552488              0.912949   

   wardTakedowns  wardTakedownsBefore20M  wardsGuarded  
0              1                       0             0  
1              3                       1             0  
2              3                       1             0  
3              4                       3             0  
4              2                       2             0  

[5 rows x 145 columns]

In [30]:
data['teamPosition'].value_counts()

teamPosition
JUNGLE     29962
TOP        29705
MIDDLE     29705
BOTTOM     29669
UTILITY    27418
Name: count, dtype: int64

In [31]:
data = data.drop(columns=['individualPosition','lane','role','playedChampSelectPosition'])

In [32]:
data.shape

(146459, 141)

In [33]:
data.isnull().sum()[lambda x : x > 0]

earlyLaningPhaseGoldExpAdvantage      914
firstTurretKilledTime               73049
killParticipation                      88
laningPhaseGoldExpAdvantage          1893
maxCsAdvantageOnLaneOpponent          791
maxLevelLeadLaneOpponent              791
visionScoreAdvantageLaneOpponent      791
dtype: int64

#### **NULL 값이 포함된 변수 설명**

* earlyLaningPhaseGoldExpAdvantage: 초반 라인전 단계에서 상대와의 비교를 통해, 얻은 골드와 경험치의 우위를 나타내는 지표

* firstTurretKilledTime: 첫 번째 포탑이 파괴된 시간(초 단위)

* killParticipation: 플레이어가 팀의 총 킬 중에 관여한 비율

* laningPhaseGoldExpAdvantage: 라인전 단계에서 골드와 경험치 우위를 나타내는 지표로, 초반과 달리 전체 라인전 기간 동안의 우위

* maxCsAdvantageOnLaneOpponent: 라인 상대와 비교했을 때 최대 CS(Creep Score) 우위

* maxLevelLeadLaneOpponent: 라인 상대와 비교했을 때 최대 레벨 우위

* visionScoreAdvantageLaneOpponent: 라인 상대와 비교했을 때 시야 점수에서의 우위.

In [ ]:
bronze = data.loc[(data['tier'] == 'BRONZE')]
silver = data.loc[(data['tier'] == 'SILVER')]
gold = data.loc[(data['tier'] == 'GOLD')]
platinum = data.loc[(data['tier'] == 'EMERALD')]
emerald = data.loc[(data['tier'] == 'PLATINUM')]
diamond = data.loc[(data['tier'] == 'DIAMOND')]
columns_to_compare = data.select_dtypes(include=['int64', 'float64']).columns.tolist()
for col in columns_to_compare:
    plt.figure(figsize=(10, 6))
    plt.hist(bronze[col], bins=20, alpha=0.5, label='BRONZE', density=True)
    plt.hist(silver[col], bins=20, alpha=0.5, label='SILVER', density=True)
    plt.hist(gold[col], bins=20, alpha=0.5, label='GOLD', density=True)
    plt.hist(platinum[col], bins=20, alpha=0.5, label='PLATINUM', density=True)
    plt.hist(emerald[col], bins=20, alpha=0.5, label='EMERALD', density=True)
    plt.hist(diamond[col], bins=20, alpha=0.5, label='DIAMOND', density=True)
    plt.xlabel(col)
    plt.ylabel('Density')
    plt.title(f'Comparison of {col} between Tier')
    plt.legend()
    plt.show()

Output hidden; open in https://colab.research.google.com to view.

* maxCsAdvantageOnLaneOpponent	791
* maxLevelLeadLaneOpponent	791
* visionScoreAdvantageLaneOpponent	791
* 특정 행에 문제가 발생했을 확률이 높고 해당 변수들은 티어별 꽤 준수?한 차이점을 보여주고 있으므로 변수를 삭제할 수 없다.

In [ ]:
data.loc[data['maxCsAdvantageOnLaneOpponent'].isnull() & data['maxLevelLeadLaneOpponent'].isnull() & data['visionScoreAdvantageLaneOpponent'].isnull()]

tier  champExperience  champLevel championName  commandPings  \
7        BRONZE             1432           4        Briar             0   
97       BRONZE             2370           5         Ryze             4   
144      BRONZE             1836           5       Ezreal             0   
233      BRONZE             3085           6       Anivia             1   
256      BRONZE             2829           6        Galio             0   
...         ...              ...         ...          ...           ...   
145836  DIAMOND             1238           4     JarvanIV             5   
146000  DIAMOND             1226           4    Seraphine             0   
146033  DIAMOND             3409           7          Jax             1   
146128  DIAMOND             2528           6   MonkeyKing             0   
146139  DIAMOND             2352           5        Garen             0   

        consumablesPurchased  damageDealtToBuildings  damageDealtToObjectives  \
7                          1                       0                      142   
97                         3                       0                        0   
144                        1                       0                        0   
233                        3                       0                        0   
256                        2                     864                      864   
...                      ...                     ...                      ...   
145836                     2                       0                        0   
146000                     3                     188                      188   
146033                     3                    2629                     2629   
146128                     2                     158                      158   
146139                     0                       0                      350   

        damageDealtToTurrets  damageSelfMitigated  ...  teamBaronKills  \
7                          0                 1340  ...               0   
97                         0                 1065  ...               0   
144                        0                  694  ...               0   
233                        0                  760  ...               0   
256                      864                 1210  ...               0   
...                      ...                  ...  ...             ...   
145836                     0                 2276  ...               0   
146000                   188                   52  ...               0   
146033                  2629                 1250  ...               0   
146128                   158                 1950  ...               0   
146139                     0                 3882  ...               0   

        teamDamagePercentage  teamElderDragonKills  teamRiftHeraldKills  \
7                   0.104461                     0                    0   
97                  0.340796                     0                    0   
144                 0.252706                     0                    0   
233                 0.385472                     0                    0   
256                 0.225822                     0                    0   
...                      ...                   ...                  ...   
145836              0.162258                     0                    0   
146000              0.356585                     0                    0   
146033              0.302547                     0                    0   
146128              0.288014                     0                    0   
146139              0.170672                     0                    0   

        turretPlatesTaken  visionScoreAdvantageLaneOpponent  \
7                       0                               NaN   
97                      0                               NaN   
144                     0                               NaN   
233                     0                               NaN   
256                     1       

In [ ]:
from IPython.display import display_html
import pandas as pd

def display_side_by_side(*args):
    """여러 데이터프레임 비교가 쉽게 옆쪽으로 표시한다"""
    html_str = ''
    for df in args:
        html_str += df.to_html()
    display_html(html_str.replace('table', 'table style="display:inline"'), raw=True)

In [ ]:
a=data['earlyLaningPhaseGoldExpAdvantage'].value_counts().reset_index()
b= data['laningPhaseGoldExpAdvantage'].value_counts().reset_index()
display_side_by_side(a,b)
# 해당 변수들에서 insight를 뽑아내기는 어려워 보인다.

earlyLaningPhaseGoldExpAdvantage 
 count 
 
 
 
 
 0 
 0.0 
 133377 
 
 
 1 
 1.0 
 12168 
 
 
 
 
 
 
 laningPhaseGoldExpAdvantage 
 count 
 
 
 
 
 0 
 0.0 
 126382 
 
 
 1 
 1.0 
 18184

* killParticipation 역시 중요 변수이므로 역시 null인 행을 삭제해야한다.
* earlyLaningPhaseGoldExpAdvantage 는 변수 자체를 삭제.
* laningPhaseGoldExpAdvantage도 변수 자체를 삭제
* firstTurretKilledTime은 데이터 분포 자체는 유용하나, null 개수가 너무 많아서 변수 드롭

In [34]:
data = data.drop(columns=['earlyLaningPhaseGoldExpAdvantage','laningPhaseGoldExpAdvantage','firstTurretKilledTime'])

In [35]:
data.shape

(146459, 138)

In [36]:
data.isnull().sum()[lambda x : x > 0]

killParticipation                    88
maxCsAdvantageOnLaneOpponent        791
maxLevelLeadLaneOpponent            791
visionScoreAdvantageLaneOpponent    791
dtype: int64

In [37]:
data.dropna(inplace=True)

In [38]:
data.isnull().sum()[lambda x : x > 0]

Series([], dtype: int64)

In [39]:
data.reset_index(drop=True,inplace=True)

In [40]:
data.shape

(145643, 138)

In [41]:
data['tier'].value_counts()

tier
DIAMOND     30377
SILVER      24062
EMERALD     23441
GOLD        22868
PLATINUM    22773
BRONZE      22122
Name: count, dtype: int64

### 지표적으로 부적합한 데이터 삭제

In [42]:
list1 = ['dangerPings',
         'dragonKills',
         'holdPings',
         'inhibitorKills',
         'item1',
         'item2',
         'item3',
         'item4',
         'item5',
         'item6',
         'objectivesStolen',
         'objectivesStolenAssists',
         'sightWardsBoughtInGame',
         'totalAllyJungleMinionsKilled',
         'turretKills',
         'turretTakedowns',
         'turretsLost',
         'visionClearedPings',
         'basicPings',
         'firstTurretKilled',
         'kTurretsDestroyedBeforePlatesFall',
         'killedChampTookFullTeamDamageSurvived',
         'takedownsInEnemyFountain',
         'teamBaronKills',
         'teamElderDragonKills',
         'teamRiftHeraldKills']

In [ ]:
for col in list1:
    print(data[col].value_counts())

dangerPings
0    145642
4         1
Name: count, dtype: int64
dragonKills
0    115593
1     16041
2      7981
3      4363
4      1510
5       150
6         5
Name: count, dtype: int64
holdPings
0    145643
Name: count, dtype: int64
inhibitorKills
0    125054
1     17261
2      2895
3       351
4        65
5        12
6         4
9         1
Name: count, dtype: int64
item1
2031    5867
0       5523
2055    4617
3157    4045
6692    3980
        ... 
2138       1
1106       1
3865       1
2140       1
3600       1
Name: count, Length: 202, dtype: int64
item2
0       9857
2055    9230
3020    5698
3047    5501
3006    4633
        ... 
2139       4
2138       4
2150       2
3400       1
3599       1
Name: count, Length: 200, dtype: int64
item3
0       9299
3020    7438
3047    6117
3158    5364
3111    4699
        ... 
2139       2
1107       1
2138       1
3865       1
2140       1
Name: count, Length: 201, dtype: int64
item4
0       19276
3020     4774
1036     4062
3158     3965
3047 

In [ ]:
data.shape

(145643, 138)

In [43]:
data = data.drop(columns=list1)

In [44]:
data.shape

(145643, 112)

In [ ]:
data.select_dtypes(include=['bool']).columns

Index(['firstBloodAssist', 'firstBloodKill', 'firstTowerAssist',
       'firstTowerKill', 'win'],
      dtype='object')

In [45]:
data = data.drop(columns=['firstBloodAssist','firstBloodKill','firstTowerAssist','firstTowerKill'])

In [46]:
data.shape

(145643, 108)

In [ ]:
data.select_dtypes(include=['object'])

tier championName teamPosition         legendaryItemUsed
0        BRONZE        Poppy          TOP              [3742, 3075]
1        BRONZE        Braum      UTILITY              [3869, 3084]
2        BRONZE      Morgana      UTILITY        [3871, 2503, 6653]
3        BRONZE       Thresh      UTILITY        [3869, 3083, 3075]
4        BRONZE        Ziggs      UTILITY  [3871, 6655, 3003, 3040]
...         ...          ...          ...                       ...
145638  DIAMOND        Leona      UTILITY                    [3877]
145639  DIAMOND        Poppy      UTILITY              [3869, 3083]
145640  DIAMOND         Zyra      UTILITY              [3871, 2503]
145641  DIAMOND       Soraka      UTILITY              [6617, 3870]
145642  DIAMOND        Leona      UTILITY  [3877, 3083, 3109, 3222]

[145643 rows x 4 columns]

In [ ]:
data['win'].value_counts()

win
True     75141
False    70502
Name: count, dtype: int64

In [ ]:
data['nexusLost'].value_counts()

nexusLost
0    102234
1     43409
Name: count, dtype: int64

In [51]:
data['gameLength'].median()

1666.353659063

In [53]:
data[(data['nexusLost']==1) & (data['gameLength'] >= 2400)]

tier  champExperience  champLevel championName  commandPings  \
44       BRONZE            19687          18         Gnar             2   
70       BRONZE            26033          18        Sivir            33   
76       BRONZE            22936          18        Fiora             0   
122      BRONZE            19562          18        Shaco             2   
128      BRONZE            15107          16       Gragas             3   
...         ...              ...         ...          ...           ...   
145362  DIAMOND            28571          18       Aatrox            22   
145385  DIAMOND            19265          18         Sett             6   
145394  DIAMOND            25905          18       Aatrox             9   
145440  DIAMOND            23236          18      Smolder             8   
145642  DIAMOND            14246          15        Leona            53   

        consumablesPurchased  damageDealtToBuildings  damageDealtToObjectives  \
44                         2                    7305                     7305   
70                         5                    4111                    17567   
76                         3                   10189                    10350   
122                        3                    2385                    15881   
128                        5                       0                     6728   
...                      ...                     ...                      ...   
145362                     8                   13086                    27330   
145385                    11                    7632                    12443   
145394                    11                    2695                     8037   
145440                     4                    3371                    13270   
145642                    15                     479                      941   

        damageDealtToTurrets  damageSelfMitigated  ...  soloKills  \
44                      7305                52416  ...          1   
70                      4111                44863  ...          1   
76                     10189                53960  ...          4   
122                     2385                36566  ...          2   
128                        0                43200  ...          1   
...                      ...                  ...  ...        ...   
145362                 13086                74414  ...          7   
145385                  7632                78354  ...          1   
145394                  2695                71090  ...          3   
145440                  3371                23669  ...          2   
145642                   479                65088  ...          0   

        stealthWardsPlaced  survivedThreeImmobilizesInFight  \
44                      13                                1   
70                      21                               10   
76                      13                                4   
122                      0                               10   
128                      0                                0   
...                    ...                              ...   
145362                   4                                0   
145385                   7                                1   
145394                   8                                2   
145440                   7                                0   
145642                  40                                5   

        teamDamagePercentage  turretPlatesTaken  \
44                  0.095469                  4   
70                  0.271689                  4   
76                  0.265525                  0   
122                 0.162304                  0   
128                 0.128197                  0   
...                      ...                ...   
145362              0.227430                  5   
145385              0.241160                  1   
145394              0.189139                  2   
145440              0.254729             

In [55]:
data[(data['nexusLost']==0) & (data['gameLength'] >= 2400)]

tier  champExperience  champLevel championName  commandPings  \
23       BRONZE            19339          18      Alistar             0   
91       BRONZE            25711          18         Ryze             9   
105      BRONZE            28174          18         Azir            13   
106      BRONZE            25346          18         Ryze            11   
115      BRONZE            19503          18        Neeko             2   
...         ...              ...         ...          ...           ...   
145372  DIAMOND            19428          18         Lulu             0   
145383  DIAMOND            17017          17         Lulu             0   
145384  DIAMOND            17662          17        Milio             1   
145459  DIAMOND            22972          18        Diana            44   
145619  DIAMOND            26540          18  TwistedFate            17   

        consumablesPurchased  damageDealtToBuildings  damageDealtToObjectives  \
23                         7                     958                     1253   
91                         6                    3244                    38865   
105                        8                    3035                    87405   
106                        9                    9282                    22010   
115                        7                     748                     8194   
...                      ...                     ...                      ...   
145372                    20                    2092                     6761   
145383                    21                     227                     3026   
145384                    16                    1944                     2890   
145459                     9                    3546                    36864   
145619                     6                   17082                    17317   

        damageDealtToTurrets  damageSelfMitigated  ...  soloKills  \
23                       958                85790  ...          0   
91                      3244                45884  ...          2   
105                     3035                27228  ...          7   
106                     9282                39917  ...          3   
115                      748                23420  ...          1   
...                      ...                  ...  ...        ...   
145372                  2092                13780  ...          1   
145383                   227                16302  ...          0   
145384                  1944                17682  ...          1   
145459                  3546                41269  ...          0   
145619                 17082                36051  ...          0   

        stealthWardsPlaced  survivedThreeImmobilizesInFight  \
23                      19                               13   
91                      18                                4   
105                     18                                7   
106                     15                                6   
115                     50                                2   
...                    ...                              ...   
145372                  39                                2   
145383                  28                                0   
145384                  38                                0   
145459                   4                                4   
145619                  13                                0   

        teamDamagePercentage  turretPlatesTaken  \
23                  0.089825                  0   
91                  0.256606                  2   
105                 0.327627                  0   
106                 0.239008                  2   
115                 0.141991                  2   
...                      ...                ...   
145372              0.081767                  4   
145383              0.045424                  0   
145384              0.054116                  2   
145459              0.241632             

In [56]:
a = data[(data['nexusLost']==1) & (data['gameLength'] >= 2400)]
b = data[(data['nexusLost']==0) & (data['gameLength'] >= 2400)]
columns_to_compare = data.select_dtypes(include=['int64', 'float64']).columns.tolist()
for col in columns_to_compare:
    plt.figure(figsize=(10, 6))
    plt.hist(a[col], bins=20, alpha=0.5, label='nexusLost & longgame', density=True)
    plt.hist(b[col], bins=20, alpha=0.5, label='win & longgame', density=True)
    plt.xlabel(col)
    plt.ylabel('Density')
    plt.title(f'Comparison of {col} nexusLost & longgame')
    plt.legend()
    plt.show()

Output hidden; open in https://colab.research.google.com to view.

In [57]:
a = data[(data['win']==1)]
b = data[(data['win']==0)]
columns_to_compare = data.select_dtypes(include=['int64', 'float64']).columns.tolist()
for col in columns_to_compare:
    plt.figure(figsize=(10, 6))
    plt.hist(a[col], bins=20, alpha=0.5, label='win', density=True)
    plt.hist(b[col], bins=20, alpha=0.5, label='loss', density=True)
    plt.xlabel(col)
    plt.ylabel('Density')
    plt.title(f'Comparison of {col} win & loss')
    plt.legend()
    plt.show()

Output hidden; open in https://colab.research.google.com to view.

* 난 항복하고 싶은데, 팀원이 항복을 안해서 게임 시간이 길어졌고 넥서스가 부셔졌다.
* 위 상황에서 인 게임 지표가 정상적일리 없다.
* 게임시간,승패변수,넥서스파괴여부를 전부 변수로 넣어서 감지할 수 있는가?

In [59]:
data['win'] = data['win'].astype('int')

In [62]:
data.shape

(145643, 108)

In [63]:
data = data.drop(columns=['championName','legendaryItemUsed'])

In [64]:
data.shape

(145643, 106)

In [65]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145643 entries, 0 to 145642
Columns: 106 entries, tier to wardsGuarded
dtypes: float64(13), int64(91), object(2)
memory usage: 117.8+ MB
